### PR TITLE
Support torch.sparse_mask on strided input with sparse CSR, CSC, BSR, and BSC mask.

### LINKED ISSUE #107370: sparse_mask method does not support masks with CSC, BSR, and BSC layouts
## Issue description

As in the title. Currently, sparse_mask supports masks with COO and CSR layouts.

## Code example

```python
>>> a = torch.tensor([[0, 1], [2, 3]])
>>> a.sparse_mask(a.to_sparse_csc())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: sparse_mask_sparse_csr expects mask to be sparse csr
```

The expected behavior is:
```python
>>> a.sparse_mask(a.to_sparse_csc())
tensor(ccol_indices=tensor([0, 1, 3]),
       row_indices=tensor([1, 0, 1]),
       values=tensor([2, 1, 3]), size=(2, 2), nnz=3, layout=torch.sparse_csc)
```

## System Info

- PyTorch version: main

cc @alexsamardzic @nikitaved @cpuhrsch @amjames @bhosmer

### LINKED ISSUE #107373: sparse_mask method does not support CSR masks with dense dimensions
## Issue description

As in the title. Currently, sparse_mask supports COO masks with dense dimensions.

## Code example

```python
>>> a = torch.tensor([[[10], [1]], [[2], [3]]])
>>> a.sparse_mask(a.to_sparse_csr())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: sparse_mask_sparse_csr expects self to be 2D
```

The expected result is
```python
>>> a.sparse_mask(a.to_sparse_csr())
tensor(crow_indices=tensor([[0, 1, 2],
                            [0, 1, 2]]),
       col_indices=tensor([[0, 0],
                           [0, 0]]),
       values=tensor([[10,  1],
                      [ 2,  3]]), size=(2, 2, 1), nnz=2,
       layout=torch.sparse_csr)
```

## System Info

- PyTorch version: main

cc @alexsamardzic @nikitaved @cpuhrsch @amjames @bhosmer

### PR DESCRIPTION
**Pull request by pearu**
While `input.sparse_mask(mask)` can be defined as `input.mul(ones_like(mask))`, implementing this definition leads to a chicken-and-egg problem because the multiplication of dense and sparse_compressed tensors relies on the `sparse_mask` support.

This PR implements `sparse_mask` support for sparse compressed masks using utility functions from sparse compressed tensor conversions support.

Fixes https://github.com/pytorch/pytorch/issues/107373
Fixes https://github.com/pytorch/pytorch/issues/107370

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107903
* #107150
* __->__ #107777
* #107638

### CODE REVIEW

File: aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp [hunk 1]
```diff
@@ -368,20 +370,64 @@ Tensor& fill_sparse_csr_(Tensor& self, const Scalar& value) {
   return unary_op_inplace(self, &TensorBase::fill_, value);
 }
 
-Tensor sparse_mask_sparse_csr(
+Tensor sparse_mask_sparse_compressed(
```
Comment (amjames) [inline aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:373]: 👍

File: aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp [hunk 2]
```diff
@@ -368,20 +370,64 @@ Tensor& fill_sparse_csr_(Tensor& self, const Scalar& value) {
...
+        [&] {
+          return at::native::dense_to_sparse_with_mask(self, dense_mask, mask.layout(), {}, mask.dense_dim());
+        },
+        [&] {
+          auto blocksize = at::sparse_csr::getBlockSize(mask);
+          return at::native::dense_to_sparse_with_mask(self, dense_mask, mask.layout(), blocksize, mask.dense_dim());
+        });
+  } else if (self.layout() == mask.layout()) {
+    // TODO: keeping this for BC but the method used here may lead to
+    // incorrect indices.
```
Comment (amjames) [inline aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:416]: This is the bug referenced by #107381 correct?

**Review by amjames — APPROVED** [review]

File: aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp (hunk 2 above)
Comment (pearu) [reply aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:416]: yes, exactly. Will need to implement an intersection kernel for sparse compressed indices to fix this.

Comment (pearu) [reply aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:416]: closing as OOS for this PR.

**Review by cpuhrsch — APPROVED** [review]

**Pull request closed without merging**